### PR TITLE
Emitting OpenSCoP from MLIR affine

### PR DESCRIPTION
**Pull request by kumasento**
This PR will create a CLI tool that emits OpenSCoP representation from MLIR affine code.

### CODE REVIEW

File: lib/EmitOpenScop/EmitOpenScop.cc [hunk 1]
```diff
@@ -0,0 +1,641 @@
...
+
+/// Insert a statement characterized by its enclosing operations into a
+/// "scattering tree". This is done by iterating through every enclosing for-op
+/// from the outermost to the innermost, and we try to traverse the tree by the
+/// IVs of these ops. If an IV does not exist, we will insert it into the tree.
+/// After that, we insert the current load/store statement into the tree as a
+/// leaf. In this progress, we keep track of all the IDs of each child we meet
+/// and the final leaf node, which will be used as the scattering.
+void insertStatement(ScatteringTreeNode *root,
+                     SmallVectorImpl<Operation *> &enclosingOps,
```
Comment (ftynse) [inline lib/EmitOpenScop/EmitOpenScop.cc:261]: Prefer ArrayRef if you don't need to change the size of the vector

File: lib/EmitOpenScop/EmitOpenScop.cc [hunk 2]
```diff
@@ -0,0 +1,641 @@
...
+                     SmallVectorImpl<unsigned> &scattering) {
+  ScatteringTreeNode *curr = root;
+
+  for (unsigned i = 0, e = enclosingOps.size(); i < e; i++) {
+    Operation *op = enclosingOps[i];
+    // We only handle for op here.
+    // TODO: is it necessary to deal with if?
```
Comment (ftynse) [inline lib/EmitOpenScop/EmitOpenScop.cc:268]: Yes

File: lib/EmitOpenScop/EmitOpenScop.cc [hunk 3]
```diff
@@ -0,0 +1,641 @@
...
+private:
+  /// Emit a OpenScop definition for a single function.
+  LogicalResult emitFuncOp(FuncOp func);
+};
+
+LogicalResult ModuleEmitter::emitFuncOp(mlir::FuncOp func) {
+  // Initialize ja new Scop per FuncOp.
+  OslScop scop;
+
+  // We iterate through every operations in the function and extrat load/store
```
Comment (ftynse) [inline lib/EmitOpenScop/EmitOpenScop.cc:485]: Typos: every operation, extract

File: lib/EmitOpenScop/EmitOpenScop.cc [hunk 4]
```diff
@@ -0,0 +1,641 @@
...
+  // operations out into loadAndStoreOps.
+  SmallVector<Operation *, 8> loadAndStoreOps;
+  func.getOperation()->walk([&](Operation *op) {
+    if (isa<AffineReadOpInterface, AffineWriteOpInterface>(op))
+      loadAndStoreOps.push_back(op);
+  });
+
+  LLVM_DEBUG(llvm::dbgs() << "Found " << Twine(loadAndStoreOps.size())
```
Comment (ftynse) [inline lib/EmitOpenScop/EmitOpenScop.cc:493]: No need for Twine here, `operator<<` accepts integers just fine.

File: lib/EmitOpenScop/EmitOpenScop.cc [hunk 5]
```diff
@@ -0,0 +1,641 @@
...
+    addAccessToScop(i, memrefId, !access.isStore(), flatExprs, domain, scop);
+  }
+
+  // Setup the context after iterated all statemenets.
+  addContextToScop(paramMap, scop);
+
+  assert(scop.validate() && "Scop created cannot be validated.");
+
+  // Print the OpenScop representation to the ostream.
+  // TODO: relate STDOUT with the os stream in the emitter.
```
Comment (ftynse) [inline lib/EmitOpenScop/EmitOpenScop.cc:610]: `llvm::outs()` gives you a `raw_ostream` that prints to STDOUT

File: lib/EmitOpenScop/OslScop.cc [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+#include "llvm/Support/raw_ostream.h"
+
+#include <vector>
+
+using namespace polymer;
+
+namespace {
+
+/// Create osl_vector from a STL vector. Since the input vector is of type
+/// int64_t, we can safely assume the osl_vector we will generate has double
```
Comment (ftynse) [inline lib/EmitOpenScop/OslScop.cc:14]: Nit: "double precision" isn't usually used to refer to int64. Precision is a floating point concept.

File: test/emit-openscop/load-store-local-vars.mlir [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+  affine.for %i = 0 to %M {
+    %0 = affine.load %A[%i] : memref<?xf32>
+    affine.store %0, %A[%i + %M] : memref<?xf32>
+  }
+
+  return
+}
+
+// CHECK: # [File generated by the OpenScop Library 0.9.2]
```
Comment (ftynse) [inline test/emit-openscop/load-store-local-vars.mlir:21]: I wouldn't check for any comments (starting with `#`), they may change or be omitted. You can keep them as C comments if necessary for readability.

File: include/polymer/OslScop.h [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+  /// Create a new relation and initialize its contents. The new relation will
+  /// be created under the scop member.
+  /// The target here is an index:
+  /// 1) if it's 0, then it means the context;
+  /// 2) otherwise, if it is a positive number, it corresponds to a statement of
+  /// id=(target-1).
+  void addRelation(int target, int type, int numRows, int numCols,
+                   int numOutputDims, int numInputDims, int numLocalDims,
+                   int numParams, std::vector<std::vector<int64_t>> &eqs,
```
[the remark this answers is not in the document]
Comment (kumasento) [reply include/polymer/OslScop.h:35]: Just refactorised this interface: using 1d vector and `ArrayRef` now.

File: include/polymer/OslScop.h [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+                   std::vector<std::vector<int64_t>> &inEqs);
```
[the remark this answers is not in the document]
Comment (kumasento) [reply include/polymer/OslScop.h:36]: Thanks! Just moved to `ArrayRef`.

File: include/polymer/OslScop.h [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+  bool validate();
+
+  /// Simply create a new statement in the linked list scop->statement.
+  void createStatement();
+
+  /// Create a new relation and initialize its contents. The new relation will
+  /// be created under the scop member.
+  /// The target here is an index:
+  /// 1) if it's 0, then it means the context;
+  /// 2) otherwise, if it is a positive number, it corresponds to a statement of
```
[the remark this answers is not in the document]
Comment (kumasento) [reply include/polymer/OslScop.h:31]: Thanks Alex, sorry I'm not sure about this approach. Is there any example?

File: lib/EmitOpenScop/EmitOpenScop.cc [hunk 11]
```diff
@@ -0,0 +1,641 @@
...
+  }
+
+  // Assuming no equality.
+  unsigned numRows = ctxInEqs.size();
+  scop.addRelation(0, OSL_TYPE_CONTEXT, numRows, numCols, 0, 0, 0, numParams,
+                   ctxEqs, ctxInEqs);
+}
+
+/// Swap the posA^th identifier with the posB^th identifier.
+/// TODO: This is from AffineStructure.h, should be turned into a publicly
```
[the remark this answers is not in the document]
Comment (kumasento) [reply lib/EmitOpenScop/EmitOpenScop.cc:134]: Sure will do!

File: lib/EmitOpenScop/EmitOpenScop.cc [hunk 12]
```diff
@@ -0,0 +1,641 @@
...
+      }
+    } else {
+      domain.addSymbolId(pos, /*id=*/param);
+    }
+  }
+}
+
+} // namespace
+
+namespace {
```
[the remark this answers is not in the document]
Comment (kumasento) [reply lib/EmitOpenScop/EmitOpenScop.cc:183]: Thanks! I just noticed [this guideline](https://llvm.org/docs/CodingStandards.html#anonymous-namespaces). Now I just place class/struct definition in anonymous namespaces and turn other functions into static.

File: lib/EmitOpenScop/EmitOpenScop.cc [hunk 13]
```diff
@@ -0,0 +1,641 @@
...
+  /// Children of the current node.
+  std::vector<std::unique_ptr<ScatteringTreeNode>> children;
+
+  /// Mapping from IV to child ID.
+  llvm::DenseMap<mlir::Value, unsigned> valueIdMap;
+
+  /// Induction variable.
+  mlir::Value iv;
+
+  /// If this node is a statement, then isLeaf is true.
```
[the remark this answers is not in the document]
Comment (kumasento) [reply lib/EmitOpenScop/EmitOpenScop.cc:249]: Thanks! I was being silly.

File: lib/EmitOpenScop/EmitOpenScop.cc (hunk 2 above)
Comment (kumasento) [reply lib/EmitOpenScop/EmitOpenScop.cc:268]: Sure, will cover if later :)

File: lib/EmitOpenScop/EmitOpenScop.cc (hunk 5 above)
Comment (kumasento) [reply lib/EmitOpenScop/EmitOpenScop.cc:610]: That's true. However, this TODO is specific to `scop.print()` actually. That function calls the `osl_scop_print` API, which only accepts `FILE *` as input. I don't think there is an easy way to do that: https://stackoverflow.com/questions/109449/getting-a-file-from-a-stdfstream

**Pull request merged**